tests/unit/test_memory_certs.py

```diff
@@ -0,0 +1,139 @@
+import socket
+import time
+
+import pytest
+from requests import RequestException, Session
+
+from nucypher.utilities.certs import (
+    Address,
+    CertificateCache,
+    P2PSession,
+    SelfSignedCertificateAdapter,
+)
+
+# Define test URLs
+VALID_URL = "https://lynx.nucypher.network:9151/status"
+INVALID_URL = "https://nonexistent-domain.com"
+
+MOCK_CERT = """-----BEGIN CERTIFICATE-----
+MIIDXTCCAkWgAwIBAgIJALm157+YvLEhMA0GCSqGSIb3DQEBCwUAMEUxCzAJBgNV
+...
+-----END CERTIFICATE-----"""
+
+
+@pytest.fixture
+def cache():
+    return CertificateCache()
+
+
+@pytest.fixture
+def adapter(cache):
+    _adapter = SelfSignedCertificateAdapter()
+    _adapter.cert_cache = cache
+    return _adapter
+
+
+@pytest.fixture
+def session(adapter):
+    s = P2PSession()
+    s.adapter = adapter  # Use the same adapter instance
+    return s
+
+
+def test_init_adapter(cache, adapter):
+    assert isinstance(adapter, SelfSignedCertificateAdapter)
+
+
+def test_cert_cache_set_get():
+    cache = CertificateCache()
+    address = Address("example.com", 443)
+    cache.set(address, MOCK_CERT)
+    assert cache.get(address) == MOCK_CERT
+
+
+def test_cert_cache_expiry():
+    cache = CertificateCache(cache_duration=1)
+    address = Address("example.com", 443)
+    cache.set(address, MOCK_CERT)
+    assert not cache.is_expired(address)
+    # Wait for the cert to expire
+    time.sleep(2)
```

Comment:
Instead of actually sleeping you can patch `time.time()` to return a time 2 seconds in the future.

Reply:
True.  Do you mind If I address this test optimization in PR #3382 in order to avoid more rebases?

Reply:
Yep no problem.